### PR TITLE
[WIP] Chameleon Clothing Refactor - Porting TG's Chameleon Clothing System

### PR DESCRIPTION
**Pull request by AzuleUtama**
**What does this PR do:**
This is a rework of the way chameleon clothing items are both stored and handled in the code.

The two existing items have been moved into a single new .dm file (code/modules/clothing/chameleon.dm). This is to keep all the chameleon clothing items in a single and easy to search for location rather than the current system where the code is in 3 different files.

The chameleon proc for glasses has also been changed to now get the replacement name, description and icons directly from the object you choose to disguise them as. This was because a fair portion of the disguises were using outdated names and/or descriptions from where the original item had changed since the chameleon HUD was added. I've left the donator gars in as an option as well since they were added in on purpose in the original PR.

As a new addition on top of this, the thermal mesons for traitors have also been given the chameleon proc used for the HUD, allowing you to disguise them so you stand out less while using them in a job that doesn't require mesons. They still have the look and name of mesons initially, so this won't change how they work for people who don't want to disguise them. The uplink description also now contains a warning about the glasses being more vulnerable to bright lights, since there is otherwise no in-game information regarding this.

**Changelog:**
:cl:
add: Thermal Imaging Glasses have been reworked into Chameleon Thermal Goggles. They can be disguised to look like other forms of eyewear by using them while in your hand.
tweak: Moved all chameleon clothing items into a single .dm file.
tweak: Refactored how the chameleon proc works for glasses. It now gets the name, description and icons directly from the object you're disguising them as.
/:cl:

### CODE REVIEW

File: code/modules/clothing/chameleon.dm [hunk 1]
```diff
@@ -1,86 +1,137 @@
...
+		for(var/U in subtypesof(/obj/item/clothing/under/color) - blocked)
+			var/obj/item/clothing/under/V = new U
+			src.clothing_choices += V
+
+		for(var/U in subtypesof(/obj/item/clothing/under/rank) - blocked)
+			var/obj/item/clothing/under/V = new U
+			src.clothing_choices += V
+		return
+
+	attackby(obj/item/clothing/under/U as obj, mob/user as mob, params)
```
Comment (Ty-Omaha) [inline code/modules/clothing/chameleon.dm]: shouldn't this be written like:

`/obj/item/clothing/under/chameleon/attackby(blahblahblah)`

?

File: code/modules/clothing/chameleon.dm [hunk 2]
```diff
@@ -1,86 +1,137 @@
...
+			to_chat(user, "<span class='warning'>Nothing happens.</span>")
+			return
+		if(istype(U, /obj/item/clothing/under))
+			if(src.clothing_choices.Find(U))
+				to_chat(user, "<span class='warning'>Pattern is already recognised by the suit.</span>")
+				return
+			src.clothing_choices += U
+			to_chat(user, "<span class='warning'>Pattern absorbed by the suit.</span>")
+
+	emp_act(severity)
```
Comment (Ty-Omaha) [inline code/modules/clothing/chameleon.dm]: same?

File: code/modules/clothing/chameleon.dm [hunk 3]
```diff
@@ -1,86 +1,137 @@
...
+		item_color = A.item_color
+		usr.update_inv_w_uniform()	//so our overlays update.
+
+/obj/item/clothing/under/chameleon/all/New()
+	..()
+	var/blocked = list(/obj/item/clothing/under/chameleon, /obj/item/clothing/under/chameleon/all)
+	//to prevent an infinite loop
+	for(var/U in typesof(/obj/item/clothing/under)-blocked)
+		var/obj/item/clothing/under/V = new U
+		src.clothing_choices += V
```
Comment (Ty-Omaha) [inline code/modules/clothing/chameleon.dm]: src is implied

File: code/modules/clothing/chameleon.dm [hunk 4]
```diff
@@ -1,86 +1,137 @@
...
+	var/list/clothing_choices = list()
+	burn_state = FIRE_PROOF
+	armor = list(melee = 10, bullet = 10, laser = 10, energy = 0, bomb = 0, bio = 0, rad = 0)
+
+	New()
+		..()
+		var/blocked = list(/obj/item/clothing/under/color/random, /obj/item/clothing/under/rank/centcom) // Stops random coloured jumpsuit and undefined centcomm suit appearing in the list.
+		for(var/U in subtypesof(/obj/item/clothing/under/color) - blocked)
+			var/obj/item/clothing/under/V = new U
+			src.clothing_choices += V
```
Comment (Ty-Omaha) [inline code/modules/clothing/chameleon.dm]: src is implied

File: code/modules/clothing/chameleon.dm [hunk 5]
```diff
@@ -1,86 +1,137 @@
...
+
+		for(var/U in subtypesof(/obj/item/clothing/under/rank) - blocked)
+			var/obj/item/clothing/under/V = new U
+			src.clothing_choices += V
```
Comment (Ty-Omaha) [inline code/modules/clothing/chameleon.dm]: src is implied

File: code/modules/clothing/chameleon.dm [hunk 6]
```diff
@@ -1,86 +1,137 @@
...
+	attackby(obj/item/clothing/under/U as obj, mob/user as mob, params)
+		..()
+		if(istype(U, /obj/item/clothing/under/chameleon))
+			to_chat(user, "<span class='warning'>Nothing happens.</span>")
+			return
+		if(istype(U, /obj/item/clothing/under))
+			if(src.clothing_choices.Find(U))
+				to_chat(user, "<span class='warning'>Pattern is already recognised by the suit.</span>")
+				return
+			src.clothing_choices += U
```
Comment (Ty-Omaha) [inline code/modules/clothing/chameleon.dm]: src is implied

File: code/modules/clothing/chameleon.dm [hunk 7]
```diff
@@ -1,86 +1,137 @@
...
+			src.clothing_choices += V
+		return
+
+	attackby(obj/item/clothing/under/U as obj, mob/user as mob, params)
+		..()
+		if(istype(U, /obj/item/clothing/under/chameleon))
+			to_chat(user, "<span class='warning'>Nothing happens.</span>")
+			return
+		if(istype(U, /obj/item/clothing/under))
+			if(src.clothing_choices.Find(U))
```
Comment (Ty-Omaha) [inline code/modules/clothing/chameleon.dm]: src is implied

**Comment by AzuleUtama** [comment]
The chameleon jumpsuit code wasn't changed other than moving it across the the new .dm file. So it's always been designed that way.

I could still change it and see if it works alright though.

**Comment by AzuleUtama** [comment]
Alright, I think I've made the chameleon suit a bit less spaghett now.

...Maybe...

...Possibly...

**Comment by datlo** [comment]
the gars are donator? i thought they were the miner gar mesons that you can buy for points, huh

**Comment by Fox-McCloud** [comment]
TG's chameleon system is a lot better than this; I'd recommend implementing that. The more icky verbs we can get rid of, the better.

**Comment by variableundefined** [comment]
Ping me when the above changes is made.

**Comment by AzuleUtama** [comment]
I'm going to close this for now while I try and work out some remaining issues with the port.

**Pull request closed without merging**